### PR TITLE
Clean dirty addresses and booleans

### PR DESCRIPTION
**Pull request by cairoeth**
Fixes M-01

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changeset entry (run `npx changeset add`)

### CODE REVIEW

File: test/utils/SlotDerivation.t.sol [hunk 1]
```diff
@@ -10,6 +10,13 @@ import {SlotDerivation} from "@openzeppelin/contracts/utils/SlotDerivation.sol";
 contract SlotDerivationTest is Test, SymTest {
     using SlotDerivation for bytes32;
 
+    function testDirtyAddress() public {
+        bytes32 result1 = SlotDerivation.deriveMapping(0, address(uint160(type(uint256).max)));
```
Comment (Amxx) [inline test/utils/SlotDerivation.t.sol:14]: This does not contain dirty upper bits. The casting to `uint160` cleans it.

File: scripts/generate/templates/SlotDerivation.js [hunk 2]
```diff
@@ -77,7 +77,13 @@ const mapping = ({ type }) => `\
  */
 function deriveMapping(bytes32 slot, ${type} key) internal pure returns (bytes32 result) {
     assembly ("memory-safe") {
-        mstore(0x00, key)
+        ${
+          type == 'address'
+            ? 'mstore(0x00, shr(96, shl(96, key)))'
```
Comment (Amxx) [inline scripts/generate/templates/SlotDerivation.js:82]: In packing we clean by doing

```suggestion
            ? 'mstore(0x00, and(key, shl(96, not(0))))'
```

I'm thinking we should an helper for cleaning

File: test/utils/SlotDerivation.t.sol (hunk 1 above)
Comment (cairoeth) [reply test/utils/SlotDerivation.t.sol:14]: only when stored in memory afaik -- https://x.com/optimizoor/status/1560590413768994816?s=46. test reverts with current master behaviour.

Comment (Amxx) [reply test/utils/SlotDerivation.t.sol:14]: Ok

I changed the tests to something that is clearer IMO, and that support dirtying any type (bool, addres, anything else we want to add)

File: scripts/generate/helpers/sanitize.js [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  address: expr => `and(${expr}, shr(96, not(0)))`,
```
Comment (ernestognw) [inline scripts/generate/helpers/sanitize.js:2]: Perhaps it makes sense to write the mask down directly. It's arguably similar in terms of readability and may shave a couple of instructions.

```suggestion
  address: expr => `and(${expr}, 0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff)`,
```

**Review by ernestognw — APPROVED** [review]
Minor nit if you're eager to discuss. LGTM

File: scripts/generate/helpers/sanitize.js (hunk 3 above)
Comment (Amxx) [reply scripts/generate/helpers/sanitize.js:2]: It increasses the code length though. Not sure which one is cheaper.

Comment (Amxx) [reply scripts/generate/helpers/sanitize.js:2]: I checked, and after optimisation both version produce the same bytecode.

Comment (Amxx) [reply scripts/generate/helpers/sanitize.js:2]: Without optimisations, the `0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff` is bigger (and more expensive to deploy), but slightly less expensive to run.

**Review by ernestognw — APPROVED** [review]

**Pull request merged**